File: packages/sdk/src/deterministic/validator.ts
import type {
  ArgumentConstraint,
  LocalValidationResult,
  ConstraintCheckResult,
} from './types.js';
import { isSafePattern } from './regex-safety.js';

export function validateDeterministic(
  _toolName: string,
  args: Record<string, unknown>,
  constraints: ArgumentConstraint[]
): LocalValidationResult {
  const startTime = Date.now();
  const validations: { argument: string; status: 'pass' | 'fail'; reason?: string }[] = [];

  for (const constraint of constraints) {
    if (!constraint.enabled) continue;

    const value = args[constraint.argumentName];

    if (value === undefined || value === null) {
      if (constraint.required) {
        return {
          decision: 'deny',
          reason: `Required argument '${constraint.argumentName}' is missing`,
          failedArgument: constraint.argumentName,
          validations,
          latencyMs: Date.now() - startTime,
        };
      }
      if (constraint.notNull && value === null) {
        return {
          decision: 'deny',
          reason: `Argument '${constraint.argumentName}' cannot be null`,
          failedArgument: constraint.argumentName,
          validations,
          latencyMs: Date.now() - startTime,
        };
      }
      continue;
    }

    const result = checkConstraints(value, constraint);

    validations.push({
      argument: constraint.argumentName,
      status: result.pass ? 'pass' : 'fail',
      reason: result.reason,
    });

    if (!result.pass) {
      return {
        decision: 'deny',
        reason: `Argument '${constraint.argumentName}' failed: ${result.reason}`,
        failedArgument: constraint.argumentName,
        validations,
        latencyMs: Date.now() - startTime,
      };
    }
  }

  return {
    decision: 'allow',
    validations,
    latencyMs: Date.now() - startTime,
  };
}

function checkConstraints(
  value: unknown,
  constraint: ArgumentConstraint
): ConstraintCheckResult {
  if (typeof value === 'number') {
    return checkNumberConstraints(value, constraint);
  }

  if (typeof value === 'string') {
    return checkStringConstraints(value, constraint);
  }

  if (Array.isArray(value)) {
    return checkArrayConstraints(value, constraint);
  }

  return { pass: true };
}

function checkNumberConstraints(
  value: number,
  constraint: ArgumentConstraint
): ConstraintCheckResult {
  if (constraint.greaterThan !== undefined && value <= constraint.greaterThan) {
    return {
      pass: false,
      reason: `value ${value} must be greater than ${constraint.greaterThan}`,
    };
  }

  if (constraint.lessThan !== undefined && value >= constraint.lessThan) {
    return {
      pass: false,
      reason: `value ${value} must be less than ${constraint.lessThan}`,
    };
  }

  if (constraint.greaterThanOrEqual !== undefined && value < constraint.greaterThanOrEqual) {
    return {
      pass: false,
      reason: `value ${value} must be >= ${constraint.greaterThanOrEqual}`,
    };
  }

  if (constraint.lessThanOrEqual !== undefined && value > constraint.lessThanOrEqual) {
    return {
      pass: false,
      reason: `value ${value} must be <= ${constraint.lessThanOrEqual}`,
    };
  }

  if (constraint.minimum !== undefined && value < constraint.minimum) {
    return {
      pass: false,
      reason: `value ${value} must be >= ${constraint.minimum}`,
    };
  }

  if (constraint.maximum !== undefined && value > constraint.maximum) {
    return {
      pass: false,
      reason: `value ${value} must be <= ${constraint.maximum}`,
    };
  }

  return { pass: true };
}

function checkStringConstraints(
  value: string,
  constraint: ArgumentConstraint
): ConstraintCheckResult {
  if (constraint.minLength !== undefined && value.length < constraint.minLength) {
    return {
      pass: false,
      reason: `length ${value.length} is less than minimum ${constraint.minLength}`,
    };
  }

  if (constraint.maxLength !== undefined && value.length > constraint.maxLength) {
    return {
      pass: false,
      reason: `length ${value.length} exceeds maximum ${constraint.maxLength}`,
    };
  }

  if (constraint.regex !== undefined) {
    if (constraint.regex.length > 256) {
      return {
        pass: false,
        reason: `regex pattern too long (${constraint.regex.length} chars, max 256)`,
      };
    }
    try {
      const regex = new RegExp(constraint.regex);
      if (!isSafePattern(constraint.regex)) {
        return {
          pass: false,
          reason: `regex pattern is potentially unsafe (ReDoS risk): ${constraint.regex}`,
        };
      }
      if (!regex.test(value)) {
        return {
          pass: false,
          reason: `value does not match pattern ${constraint.regex}`,
        };
      }
    } catch {
      return {
        pass: false,
        reason: `invalid regex pattern: ${constraint.regex}`,
      };
    }
  }

  if (constraint.enum !== undefined && !constraint.enum.includes(value)) {
    return {
      pass: false,
      reason: `value "${value}" is not in allowed values: ${constraint.enum.join(', ')}`,
    };
  }

  return { pass: true };
}

function checkArrayConstraints(
  value: unknown[],
  constraint: ArgumentConstraint
): ConstraintCheckResult {
  if (constraint.minItems !== undefined && value.length < constraint.minItems) {
    return {
      pass: false,
      reason: `array has ${value.length} items, minimum is ${constraint.minItems}`,
    };
  }

  if (constraint.maxItems !== undefined && value.length > constraint.maxItems) {
    return {
      pass: false,
      reason: `array has ${value.length} items, maximum is ${constraint.maxItems}`,
    };
  }

  return { pass: true };
}

File: packages/sdk-python/veto/deterministic/validator.py
import re
import time
from typing import Any

from veto.deterministic.types import (
    ArgumentConstraint,
    ConstraintCheckResult,
    LocalValidationResult,
    ValidationEntry,
)
from veto.deterministic.regex_safety import is_safe_pattern


def validate_deterministic(
    tool_name: str,
    args: dict[str, Any],
    constraints: list[ArgumentConstraint],
) -> LocalValidationResult:
    start = time.monotonic()
    validations: list[ValidationEntry] = []

    for constraint in constraints:
        if not constraint.enabled:
            continue

        value = args.get(constraint.argument_name)

        if value is None:
            key_exists = constraint.argument_name in args

            if constraint.required and not key_exists:
                return LocalValidationResult(
                    decision="deny",
                    reason=f"Required argument '{constraint.argument_name}' is missing",
                    failed_argument=constraint.argument_name,
                    validations=validations,
                    latency_ms=_elapsed_ms(start),
                )
            if constraint.not_null and key_exists:
                return LocalValidationResult(
                    decision="deny",
                    reason=f"Argument '{constraint.argument_name}' cannot be null",
                    failed_argument=constraint.argument_name,
                    validations=validations,
                    latency_ms=_elapsed_ms(start),
                )
            continue

        result = _check_constraints(value, constraint)

        validations.append(
            ValidationEntry(
                argument=constraint.argument_name,
                status="pass" if result.passed else "fail",
                reason=result.reason,
            )
        )

        if not result.passed:
            return LocalValidationResult(
                decision="deny",
                reason=f"Argument '{constraint.argument_name}' failed: {result.reason}",
                failed_argument=constraint.argument_name,
                validations=validations,
                latency_ms=_elapsed_ms(start),
            )

    return LocalValidationResult(
        decision="allow",
        validations=validations,
        latency_ms=_elapsed_ms(start),
    )


def _elapsed_ms(start: float) -> float:
    return (time.monotonic() - start) * 1000


def _check_constraints(value: Any, constraint: ArgumentConstraint) -> ConstraintCheckResult:
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return _check_number_constraints(value, constraint)
    if isinstance(value, str):
        return _check_string_constraints(value, constraint)
    if isinstance(value, list):
        return _check_array_constraints(value, constraint)
    return ConstraintCheckResult(passed=True)


def _check_number_constraints(
    value: float, constraint: ArgumentConstraint
) -> ConstraintCheckResult:
    import math

    if math.isnan(value):
        return ConstraintCheckResult(passed=False, reason="value is NaN")

    if math.isinf(value):
        return ConstraintCheckResult(
            passed=False, reason=f"value {value} is not finite"
        )

    if constraint.greater_than is not None and value <= constraint.greater_than:
        return ConstraintCheckResult(
            passed=False,
            reason=f"value {value} must be greater than {constraint.greater_than}",
        )
    if constraint.less_than is not None and value >= constraint.less_than:
        return ConstraintCheckResult(
            passed=False,
            reason=f"value {value} must be less than {constraint.less_than}",
        )
    if constraint.greater_than_or_equal is not None and value < constraint.greater_than_or_equal:
        return ConstraintCheckResult(
            passed=False,
            reason=f"value {value} must be >= {constraint.greater_than_or_equal}",
        )
    if constraint.less_than_or_equal is not None and value > constraint.less_than_or_equal:
        return ConstraintCheckResult(
            passed=False,
            reason=f"value {value} must be <= {constraint.less_than_or_equal}",
        )
    if constraint.minimum is not None and value < constraint.minimum:
        return ConstraintCheckResult(
            passed=False,
            reason=f"value {value} must be >= {constraint.minimum}",
        )
    if constraint.maximum is not None and value > constraint.maximum:
        return ConstraintCheckResult(
            passed=False,
            reason=f"value {value} must be <= {constraint.maximum}",
        )
    return ConstraintCheckResult(passed=True)


def _check_string_constraints(
    value: str, constraint: ArgumentConstraint
) -> ConstraintCheckResult:
    if constraint.min_length is not None and len(value) < constraint.min_length:
        return ConstraintCheckResult(
            passed=False,
            reason=f"length {len(value)} is less than minimum {constraint.min_length}",
        )
    if constraint.max_length is not None and len(value) > constraint.max_length:
        return ConstraintCheckResult(
            passed=False,
            reason=f"length {len(value)} exceeds maximum {constraint.max_length}",
        )
    if constraint.regex is not None:
        if len(constraint.regex) > 256:
            return ConstraintCheckResult(
                passed=False,
                reason=f"regex pattern too long ({len(constraint.regex)} chars, max 256)",
            )
        try:
            if not is_safe_pattern(constraint.regex):
                return ConstraintCheckResult(
                    passed=False,
                    reason=f"regex pattern is potentially unsafe (ReDoS risk): {constraint.regex}",
                )
            compiled = re.compile(constraint.regex)
            if not compiled.search(value):
                return ConstraintCheckResult(
                    passed=False,
                    reason=f"value does not match pattern {constraint.regex}",
                )
        except re.error:
            return ConstraintCheckResult(
                passed=False,
                reason=f"invalid regex pattern: {constraint.regex}",
            )
    if constraint.enum is not None and value not in constraint.enum:
        return ConstraintCheckResult(
            passed=False,
            reason=f'value "{value}" is not in allowed values: {", ".join(constraint.enum)}',
        )
    return ConstraintCheckResult(passed=True)


def _check_array_constraints(
    value: list[Any], constraint: ArgumentConstraint
) -> ConstraintCheckResult:
    if constraint.min_items is not None and len(value) < constraint.min_items:
        return ConstraintCheckResult(
            passed=False,
            reason=f"array has {len(value)} items, minimum is {constraint.min_items}",
        )
    if constraint.max_items is not None and len(value) > constraint.max_items:
        return ConstraintCheckResult(
            passed=False,
            reason=f"array has {len(value)} items, maximum is {constraint.max_items}",
        )
    return ConstraintCheckResult(passed=True)
